UBSAN failures in test/unit/bitmap and test/unit/hash

I'm looking into another bug (probable miscompilation with development version of a compiler, unrelated to this bug) and wanted to try UBSAN. I hit a failure in the jemalloc testsuite:

Using GCC 14 to avoid the other issue I'm investigating:
```
$ git clone https://github.com/jemalloc/jemalloc && cd jemalloc # at commit 17881ebbfd76529904e826f425f3266834cf3a75
$ export UBSAN_OPTIONS="halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1"
$ ./configure CC=gcc-14 CXX=g++-14 CFLAGS="-O2 -ggdb3 -fsanitize=undefined" CXXFLAGS="-O2 -ggdb3 -fsanitize=undefined" LDFLAGS="-fsanitize=undefined"
[...]
$ make && make check
=== test/unit/bitmap ===
test_bitmap_initializer (non-reentrant): pass
test_bitmap_initializer (libc-reentrant): pass
test_bitmap_initializer (arena_new-reentrant): pass
test_bitmap_size (non-reentrant): pass
test_bitmap_size (libc-reentrant): pass
test_bitmap_size (arena_new-reentrant): pass
test_bitmap_init (non-reentrant): pass
test_bitmap_init (libc-reentrant): pass
test_bitmap_init (arena_new-reentrant): pass
test_bitmap_set (non-reentrant): pass
test_bitmap_set (libc-reentrant): pass
test_bitmap_set (arena_new-reentrant): pass
test_bitmap_unset (non-reentrant): pass
test_bitmap_unset (libc-reentrant): pass
test_bitmap_unset (arena_new-reentrant): pass
include/jemalloc/internal/bitmap.h:293:5: runtime error: load of address 0x744d1b608008 with insufficient space for an object of type 'const bitmap_t'
0x744d1b608008: note: pointer points here
 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^
    #0 0x60b99a510d74 in bitmap_ffu include/jemalloc/internal/bitmap.h:293
    #1 0x60b99a510d74 in test_bitmap_xfu_body test/unit/bitmap.c:275
    #2 0x60b99a512b87 in test_bitmap_xfu test/unit/bitmap.c:323
    #3 0x60b99a740b81 in p_test_impl test/src/test.c:149
    #4 0x60b99a7411e5 in p_test test/src/test.c:200
    #5 0x60b99a504e26 in main test/unit/bitmap.c:336
    #6 0x744d1ba05746 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #7 0x744d1ba057f6 in __libc_start_main_impl ../csu/libc-start.c:360
    #8 0x60b99a504e50 in _start (/home/sam/git/jemalloc/test/unit/bitmap+0x1be50)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior include/jemalloc/internal/bitmap.h:293:5
Aborted (core dumped)
Test harness error: test/unit/bitmap w/ MALLOC_CONF=""
Use prefix to debug, e.g. JEMALLOC_TEST_PREFIX="gdb --args" sh test/test.sh test/unit/bitmap
make: *** [Makefile:699: check_unit] Error 1
```

--- a/include/jemalloc/internal/bitmap.h
+++ b/include/jemalloc/internal/bitmap.h
@@ -284,14 +284,17 @@ bitmap_ffu(const bitmap_t *bitmap, const bitmap_info_t *binfo, size_t min_bit) {
 	bitmap_t g = bitmap[i] & ~((1LU << (min_bit & BITMAP_GROUP_NBITS_MASK))
 	    - 1);
 	size_t bit;
-	do {
+	while (1) {
 		if (g != 0) {
 			bit = ffs_lu(g);
 			return (i << LG_BITMAP_GROUP_NBITS) + bit;
 		}
 		i++;
+		if (i >= binfo->ngroups) {
+			break;
+		}
 		g = bitmap[i];
-	} while (i < binfo->ngroups);
+	}
 	return binfo->nbits;
 #endif
 }

